onnxruntime/core/util/thread_utils.cc

```diff
@@ -7,9 +7,109 @@
+                        << last_error
+                        << " error msg: " << std::system_category().message(last_error);
+    return {};
+  }
+
+  std::unique_ptr<char[]> allocation = std::make_unique<char[]>(returnLength);
+  SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX* processorInfos = reinterpret_cast<SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX*>(allocation.get());
+
+  if (!GetLogicalProcessorInformationEx(RelationGroup, processorInfos, &returnLength)) {
+    last_error = GetLastError();
+    LOGS_DEFAULT(ERROR) << "GetLogicalProcessorInformationEx failed to obtain processor info. error code: "
+                        << last_error
+                        << " error msg: " << std::system_category().message(last_error);
+    return {};
+  }
+
+  WORD group_id = 0;
+  static constexpr WORD max_group_id = std::numeric_limits<WORD>::max();
+  ThreadAffinities thread_affinities;
+  // size of KAFFINITY varies, must not exceed boundary
+  int64_t num_bit_in_affinity_ = sizeof(KAFFINITY) << 3;
+  // allow only one thread for every two logical processors
+  int64_t max_num_thread_per_group = num_bit_in_affinity_ >> 1;
+
+  auto num_processor_info = returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX);
+  for (int64_t i = 0; i < static_cast<int64_t>(num_processor_info); ++i) {
+    if (processorInfos[i].Relationship != RelationGroup) {
+      LOGS_DEFAULT(ERROR) << "Returned processors not belong to same group";
+      return {};
+    }
+    for (int64_t j = 0; j < static_cast<int>(processorInfos[i].Group.ActiveGroupCount); ++j) {
+      const auto& groupInfo = processorInfos[i].Group.GroupInfo[j];
+      LOGS_DEFAULT(INFO) << "Discovered process group: " << group_id << ", active processor mask: " << groupInfo.ActiveProcessorMask;
+      KAFFINITY thread_affinity = 3UL;
+      // allow only one thread for every two logical processors
+      int64_t num_of_physical_cores = groupInfo.ActiveProcessorCount >> 1;
+      int64_t max_num_thread_cur_group = std::min( max_num_thread_per_group, num_of_physical_cores);
+      for (int64_t k = 0; k < max_num_thread_cur_group; ++k) {
+        thread_affinities.push_back({static_cast<int64_t>(group_id), static_cast<int64_t>(thread_affinity)});
+        thread_affinity <<= 2;
+      }
+      if (group_id == max_group_id) {
+        // preventing overflow
+        return std::move(thread_affinities);
+      }
+      group_id++;
+    }
+  }
+  return std::move(thread_affinities);
+}
+#endif
+
 namespace concurrency {
+
+bool ExtractAffinityFromString(const char* affinity_string, ThreadAffinities& group_affinities) {
+  group_affinities.clear();
+  auto Split = [](const std::string& s, char splitor) {
+    std::vector<std::string> ans;
+    std::string tmp;
+    std::stringstream ss;
+    ss << s;
+    while (getline(ss, tmp, splitor)) {
+      if (!tmp.empty()) {
+        ans.push_back(tmp);
+      }
+    }
+    return ans;
+  };
```

Comment:
This parsing code is a bit fragile: it can be thrown off by whitespace in the strings, e.g "1, 10" rather than "1,10". Perhaps accepting an array of integers as suggested by another reviewer would make this more robust?

Reply:
Space does not hurt stoll. But assume it does, how do we separate "an array integers"?

Reply:
I'm not sure I follow: what's to separate if your API expects an array of integers, rather than a string that represents an array of ints?

Reply:
> I'm not sure I follow: what's to separate if your API expects an array (or two arrays) of integers, rather than a string that represents an array (resp. 2 arrays) of ints?

I got your point  - the problem is "array of integers" approach brings way more ambiguity and challenges as listed in another comment.